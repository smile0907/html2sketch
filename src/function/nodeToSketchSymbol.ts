import { SymbolMaster } from '../model';
import { GroupLayoutType } from '../helpers/layout';
import nodeToSketchGroup from './nodeToSketchGroup';

interface NodeToSketchSymbolOptions {
  symbolLayout?: GroupLayoutType;
  /**
   * 如果需要对 symbol 进行调整处理
   * 传入这个方法
   */
  handleSymbol?: (symbol: SymbolMaster) => void;
}

/**
 * 解析为 Symbol
 */

export default (
  node: Element,
  { symbolLayout, handleSymbol }: NodeToSketchSymbolOptions,
) => {
  const group = nodeToSketchGroup(node);

  const symbol = new SymbolMaster({
    x: group.x,
    y: group.y,
    width: group.width,
    height: group.height,
  });

  symbol.style = group.style;

  group.layers.forEach((layer) => {
    switch (layer.class) {
      case 'text':
        // 对所有的文本都添加
        symbol.addOverride(layer.id, 'text');
        break;
      case 'svg':
        // 由于svg 属于 ShapeGroup 其坐标关系不一致因此需要重新定义过
        layer.layers.forEach((shapeGroup) => {
          shapeGroup.x += group.x;
          shapeGroup.y += group.y;
        });
        break;
      default:
    }
    symbol.layers.push(layer);
  });

  if (symbolLayout) {
    symbol.setGroupLayout(symbolLayout);
  }

  if (handleSymbol) {
    handleSymbol(symbol);
  }
  return symbol;
};
