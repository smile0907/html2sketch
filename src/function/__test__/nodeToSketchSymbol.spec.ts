import nodeToSketchSymbol from '../nodeToSketchSymbol';

describe('nodeToSketchSymbol', function () {
  afterAll(() => {
    document.body.innerHTML = '';
  });
  beforeAll(() => {
    document.body.innerHTML = `<div id="group">123</div>`;
  });
  it('单个正常解析', function () {
    const node = document.getElementById('group') as HTMLDivElement;

    const group = nodeToSketchSymbol(node);
    expect(group.toSketchJSON()).toMatchSnapshot();
  });
});
