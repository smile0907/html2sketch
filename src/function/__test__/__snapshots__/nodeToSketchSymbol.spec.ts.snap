// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeToSketchSymbol 单个正常解析 1`] = `
Object {
  "_class": "symbolMaster",
  "allowsOverrides": true,
  "backgroundColor": Object {
    "_class": "color",
    "alpha": 1,
    "blue": 1,
    "green": 1,
    "red": 1,
  },
  "booleanOperation": -1,
  "changeIdentifier": 0,
  "do_objectID": "UUID",
  "exportOptions": Object {
    "_class": "exportOptions",
    "exportFormats": Array [],
    "includedLayerIds": Array [],
    "layerOptions": 0,
    "shouldTrim": false,
  },
  "frame": Object {
    "_class": "rect",
    "constrainProportions": false,
    "height": 22,
    "width": 25.609375,
    "x": 8,
    "y": 8,
  },
  "groupLayout": Object {
    "_class": "MSImmutableFreeformGroupLayout",
  },
  "hasBackgroundColor": false,
  "hasClickThrough": true,
  "horizontalRulerData": Object {
    "_class": "rulerData",
    "base": 0,
    "guides": Array [],
  },
  "includeBackgroundColorInExport": true,
  "includeBackgroundColorInInstance": false,
  "includeInCloudUpload": true,
  "isFixedToViewport": false,
  "isFlippedHorizontal": false,
  "isFlippedVertical": false,
  "isFlowHome": false,
  "isLocked": false,
  "isVisible": true,
  "layerListExpandedType": 0,
  "layers": Array [
    Object {
      "_class": "text",
      "attributedString": Object {
        "_class": "attributedString",
        "attributes": Array [
          Object {
            "_class": "stringAttribute",
            "attributes": Object {
              "MSAttributedStringColorAttribute": Object {
                "_class": "color",
                "alpha": 1,
                "blue": 0,
                "green": 0,
                "red": 0,
              },
              "MSAttributedStringFontAttribute": Object {
                "_class": "fontDescriptor",
                "attributes": Object {
                  "name": "PingFangSC-Regular",
                  "size": 16,
                },
              },
              "MSAttributedStringTextTransformAttribute": 0,
              "kerning": 0,
              "paragraphStyle": Object {
                "_class": "paragraphStyle",
                "alignment": 0,
                "maximumLineHeight": 22,
                "minimumLineHeight": 22,
              },
              "strikethroughStyle": 0,
              "underlineStyle": 0,
            },
            "length": 3,
            "location": 0,
          },
        ],
        "string": "123",
      },
      "automaticallyDrawOnUnderlyingPath": false,
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "UUID",
      "dontSynchroniseWithSymbol": false,
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": false,
        "height": 22,
        "width": 25.609375,
        "x": 0,
        "y": 0,
      },
      "glyphBounds": "",
      "hasClippingMask": false,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "lineSpacingBehaviour": 2,
      "name": "123",
      "nameIsFixed": false,
      "resizingConstraint": 63,
      "resizingType": 0,
      "rotation": 0,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "UUID",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
      "textBehaviour": 0,
      "userInfo": undefined,
    },
  ],
  "name": "123",
  "nameIsFixed": false,
  "overrideProperties": Array [],
  "resizesContent": false,
  "resizingConstraint": 1,
  "resizingType": 1,
  "rotation": 0,
  "sharedStyleID": "",
  "shouldBreakMaskChain": false,
  "symbolID": "UUID",
  "verticalRulerData": Object {
    "_class": "rulerData",
    "base": 0,
    "guides": Array [],
  },
}
`;
